File: dhrms-backend/src/main/java/com/dhrms/dhrms_backend/service/WorkerService.java
package com.dhrms.dhrms_backend.service;

import java.util.List;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.dhrms.dhrms_backend.dto.worker.CreateWorkerRequest;
import com.dhrms.dhrms_backend.dto.worker.UpdateWorkerRequest;
import com.dhrms.dhrms_backend.dto.worker.WorkerMedicalRecordResponse;
import com.dhrms.dhrms_backend.dto.worker.WorkerProfileResponse;
import com.dhrms.dhrms_backend.dto.worker.WorkerResponse;
import com.dhrms.dhrms_backend.entity.Doctor;
import com.dhrms.dhrms_backend.entity.Hospital;
import com.dhrms.dhrms_backend.entity.User;
import com.dhrms.dhrms_backend.entity.Worker;
import com.dhrms.dhrms_backend.entity.enums.UserRole;
import com.dhrms.dhrms_backend.entity.enums.UserStatus;
import com.dhrms.dhrms_backend.repository.DoctorWorkerAssignmentRepository;
import com.dhrms.dhrms_backend.repository.HospitalRepository;
import com.dhrms.dhrms_backend.repository.MedicalRecordRepository;
import com.dhrms.dhrms_backend.repository.UserRepository;
import com.dhrms.dhrms_backend.repository.WorkerRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class WorkerService {

        private final UserRepository userRepository;

        private final PasswordEncoder passwordEncoder;

        private final WorkerRepository workerRepository;

        private final HospitalRepository hospitalRepository;

        private final DoctorWorkerAssignmentRepository assignmentRepository;
        private final MedicalRecordRepository medicalRecordRepository;

        @Transactional
        public WorkerResponse createWorker(
                        Long hospitalUserId,
                        CreateWorkerRequest request) {

                /*
                 * Verify that the authenticated user
                 * actually belongs to a hospital.
                 */
                getHospitalByUserId(hospitalUserId);

                /*
                 * Email must be unique because it is used
                 * for authentication.
                 */
                if (userRepository.findByEmail(request.getEmail()).isPresent()) {
                        throw new IllegalArgumentException(
                                        "Email is already registered");
                }

                /*
                 * Create authentication account.
                 */
                User user = User.builder()
                                .email(request.getEmail())
                                .passwordHash(
                                                passwordEncoder.encode(
                                                                request.getPassword()))
                                .role(UserRole.WORKER)
                                .status(UserStatus.ACTIVE)
                                .build();

                user = userRepository.save(user);

                /*
                 * Create worker profile.
                 */
                Worker worker = Worker.builder()
                                .user(user)
                                .workerCode(generateWorkerCode())
                                .fullName(request.getFullName())
                                .dateOfBirth(request.getDateOfBirth())
                                .gender(request.getGender())
                                .bloodGroup(request.getBloodGroup())
                                .phone(request.getPhone())
                                .address(request.getAddress())
                                .emergencyContactName(
                                                request.getEmergencyContactName())
                                .emergencyContactPhone(
                                                request.getEmergencyContactPhone())
                                .emergencyContactRelation(
                                                request.getEmergencyContactRelation())
                                .active(true)
                                .build();

                worker = workerRepository.save(worker);

                return toResponse(worker);
        }

        @Transactional(readOnly = true)
        public List<WorkerResponse> getWorkers(
                        Long hospitalUserId) {

                getHospitalByUserId(hospitalUserId);

                return workerRepository
                                .findByActiveTrue()
                                .stream()
                                .map(this::toResponse)
                                .toList();
        }

        @Transactional(readOnly = true)
        public WorkerResponse getWorker(
                        Long hospitalUserId,
                        Long workerId) {

                Hospital hospital = getHospitalByUserId(hospitalUserId);

                Worker worker = workerRepository
                                .findById(workerId)
                                .orElseThrow(() -> new IllegalArgumentException(
                                                "Worker not found"));

                WorkerResponse.WorkerResponseBuilder response = WorkerResponse.builder()
                                .id(worker.getId())
                                .workerCode(worker.getWorkerCode())
                                .fullName(worker.getFullName())
                                .dateOfBirth(worker.getDateOfBirth())
                                .gender(worker.getGender())
                                .bloodGroup(worker.getBloodGroup())
                                .phone(worker.getPhone())
                                .address(worker.getAddress())
                                .emergencyContactName(
                                                worker.getEmergencyContactName())
                                .emergencyContactPhone(
                                                worker.getEmergencyContactPhone())
                                .emergencyContactRelation(
                                                worker.getEmergencyContactRelation())
                                .active(worker.getActive())
                                .createdAt(worker.getCreatedAt());

                assignmentRepository
                                .findByWorker_IdAndHospital_IdAndActiveTrue(
                                                workerId,
                                                hospital.getId())
                                .ifPresent(assignment -> {

                                        Doctor doctor = assignment.getDoctor();

                                        response.assignedDoctorId(
                                                        doctor.getId());

                                        response.assignedDoctorName(
                                                        doctor.getFullName());

                                        response.assignedDoctorSpecialization(
                                                        doctor.getSpecialization());
                                });

                return response.build();
        }

        @Transactional(readOnly = true)
        public WorkerResponse getWorkerByCode(
                        Long hospitalUserId,
                        String workerCode) {

                getHospitalByUserId(hospitalUserId);

                Worker worker = workerRepository
                                .findByWorkerCode(workerCode)
                                .orElseThrow(() -> new IllegalArgumentException(
                                                "Worker not found"));

                return toResponse(worker);
        }

        @Transactional(readOnly = true)
        public WorkerResponse getWorkerForDoctor(
                        Long workerId) {

                Worker worker = workerRepository
                                .findById(workerId)
                                .orElseThrow(() -> new IllegalArgumentException(
                                                "Worker not found"));

                return toResponse(worker);
        }

        @Transactional
        public WorkerResponse updateWorker(
                        Long hospitalUserId,
                        Long workerId,
                        UpdateWorkerRequest request) {

                getHospitalByUserId(hospitalUserId);

                Worker worker = workerRepository
                                .findById(workerId)
                                .orElseThrow(() -> new IllegalArgumentException(
                                                "Worker not found"));

                if (request.getFullName() != null) {
                        worker.setFullName(
                                        request.getFullName());
                }

                if (request.getDateOfBirth() != null) {
                        worker.setDateOfBirth(
                                        request.getDateOfBirth());
                }

                if (request.getGender() != null) {
                        worker.setGender(
                                        request.getGender());
                }

                if (request.getBloodGroup() != null) {
                        worker.setBloodGroup(
                                        request.getBloodGroup());
                }

                if (request.getPhone() != null) {
                        worker.setPhone(
                                        request.getPhone());
                }

                if (request.getAddress() != null) {
                        worker.setAddress(
                                        request.getAddress());
                }

                if (request.getEmergencyContactName() != null) {
                        worker.setEmergencyContactName(
                                        request.getEmergencyContactName());
                }

                if (request.getEmergencyContactPhone() != null) {
                        worker.setEmergencyContactPhone(
                                        request.getEmergencyContactPhone());
                }

                if (request.getEmergencyContactRelation() != null) {
                        worker.setEmergencyContactRelation(
                                        request.getEmergencyContactRelation());
                }

                return toResponse(
                                workerRepository.save(worker));
        }

        @Transactional
        public WorkerResponse deactivateWorker(
                        Long hospitalUserId,
                        Long workerId) {

                getHospitalByUserId(hospitalUserId);

                Worker worker = workerRepository
                                .findById(workerId)
                                .orElseThrow(() -> new IllegalArgumentException(
                                                "Worker not found"));

                worker.setActive(false);

                return toResponse(
                                workerRepository.save(worker));
        }

        @Transactional
        public WorkerResponse activateWorker(
                        Long hospitalUserId,
                        Long workerId) {

                getHospitalByUserId(hospitalUserId);

                Worker worker = workerRepository
                                .findById(workerId)
                                .orElseThrow(() -> new IllegalArgumentException(
                                                "Worker not found"));

                worker.setActive(true);

                return toResponse(
                                workerRepository.save(worker));
        }

        private Hospital getHospitalByUserId(
                        Long userId) {

                return hospitalRepository
                                .findByUser_Id(userId)
                                .orElseThrow(() -> new IllegalArgumentException(
                                                "Hospital profile not found"));
        }

        private String generateWorkerCode() {

                String code;

                do {
                        code = "DHRMS-WKR-" +
                                        String.format(
                                                        "%08d",
                                                        (long) (Math.random() * 100000000));

                } while (workerRepository.existsByWorkerCode(code));

                return code;
        }

        private WorkerResponse toResponse(
                        Worker worker) {

                return WorkerResponse.builder()
                                .id(worker.getId())
                                .workerCode(worker.getWorkerCode())
                                .fullName(worker.getFullName())
                                .dateOfBirth(worker.getDateOfBirth())
                                .gender(worker.getGender())
                                .bloodGroup(worker.getBloodGroup())
                                .phone(worker.getPhone())
                                .address(worker.getAddress())
                                .emergencyContactName(
                                                worker.getEmergencyContactName())
                                .emergencyContactPhone(
                                                worker.getEmergencyContactPhone())
                                .emergencyContactRelation(
                                                worker.getEmergencyContactRelation())
                                .active(worker.getActive())
                                .createdAt(worker.getCreatedAt())
                                .build();
        }

        public WorkerProfileResponse getMyProfile(Long userId) {

                Worker worker = workerRepository.findByUserId(userId)
                                .orElseThrow(() -> new RuntimeException("Worker profile not found"));

                return WorkerProfileResponse.builder()
                                .id(worker.getId())
                                .workerCode(worker.getWorkerCode())
                                .fullName(worker.getFullName())
                                .dateOfBirth(worker.getDateOfBirth())
                                .gender(worker.getGender())
                                .bloodGroup(worker.getBloodGroup())
                                .phone(worker.getPhone())
                                .address(worker.getAddress())
                                .emergencyContactName(worker.getEmergencyContactName())
                                .emergencyContactPhone(worker.getEmergencyContactPhone())
                                .emergencyContactRelation(worker.getEmergencyContactRelation())
                                .build();
        }

        public List<WorkerMedicalRecordResponse> getMyMedicalRecords(Long userId) {

                Worker worker = workerRepository.findByUserId(userId)
                                .orElseThrow(() -> new RuntimeException("Worker profile not found"));

                return medicalRecordRepository
                                .findByWorkerIdOrderByVisitDateDesc(worker.getId())
                                .stream()
                                .map(record -> WorkerMedicalRecordResponse.builder()
                                                .id(record.getId())
                                                .visitDate(record.getVisitDate())
                                                .hospitalName(
                                                                record.getHospital().getName())
                                                .doctorName(
                                                                record.getDoctor().getFullName())
                                                .symptoms(record.getSymptoms())
                                                .diagnosis(record.getDiagnosis())
                                                .treatment(record.getTreatment())
                                                .notes(record.getNotes())
                                                .build())
                                .toList();
        }
}

File: dhrms-backend-nestjs/src/worker/worker.service.ts
import { BadRequestException, Injectable, NotFoundException, ForbiddenException } from '@nestjs/common';
import * as bcrypt from 'bcrypt';
import { PrismaService } from '../prisma/prisma.service';
import { CreateWorkerDto } from './dto/create-worker.dto';
import { UpdateWorkerDto } from './dto/update-worker.dto';

@Injectable()
export class WorkerService {
  constructor(private readonly prisma: PrismaService) {}

  private async getHospitalByUserId(userId: bigint) {
    const hospital = await this.prisma.hospital.findUnique({ where: { userId } });
    if (!hospital) throw new NotFoundException('Hospital profile not found');
    if (hospital.status !== 'ACTIVE') throw new ForbiddenException('Hospital account is not active');
    return hospital;
  }

  private async getOwnedWorker(hospitalUserId: bigint, workerId: bigint) {
    const hospital = await this.getHospitalByUserId(hospitalUserId);
    const worker = await this.prisma.worker.findFirst({ where: { id: workerId, hospitalId: hospital.id } });
    if (!worker) throw new NotFoundException('Worker not found in this hospital');
    return { hospital, worker };
  }

  async createWorker(hospitalUserId: bigint, request: CreateWorkerDto) {
    const hospital = await this.getHospitalByUserId(hospitalUserId);
    const existing = await this.prisma.user.findUnique({ where: { email: request.email } });
    if (existing) throw new BadRequestException('Email is already registered');

    const result = await this.prisma.$transaction(async (tx) => {
      const user = await tx.user.create({
        data: { email: request.email, passwordHash: await bcrypt.hash(request.password, 10), role: 'WORKER', status: 'ACTIVE' },
      });

      let workerCode: string;
      do {
        workerCode = `DHRMS-WKR-${Math.floor(Math.random() * 100000000).toString().padStart(8, '0')}`;
      } while (await tx.worker.findUnique({ where: { workerCode } }));

      const worker = await tx.worker.create({
        data: {
          userId: user.id,
          hospitalId: hospital.id,
          workerCode,
          fullName: request.fullName,
          dateOfBirth: request.dateOfBirth ? new Date(request.dateOfBirth) : undefined,
          gender: request.gender,
          bloodGroup: request.bloodGroup,
          phone: request.phone,
          address: request.address,
          emergencyContactName: request.emergencyContactName,
          emergencyContactPhone: request.emergencyContactPhone,
          emergencyContactRelation: request.emergencyContactRelation,
          active: true,
        },
      });
      return worker;
    });

    return { ...this.toResponse(result), hospitalId: Number(hospital.id), assignedDoctor: null };
  }

  async getWorkers(hospitalUserId: bigint) {
    const hospital = await this.getHospitalByUserId(hospitalUserId);
    const workers = await this.prisma.worker.findMany({
      where: { hospitalId: hospital.id },
      include: { assignments: { where: { hospitalId: hospital.id, active: true }, include: { doctor: true }, take: 1 } },
      orderBy: { createdAt: 'desc' },
    });
    return workers.map((worker) => this.withAssignment(this.toResponse(worker), worker.assignments[0]));
  }

  async getWorker(hospitalUserId: bigint, workerId: bigint) {
    const { worker } = await this.getOwnedWorker(hospitalUserId, workerId);
    const assignment = await this.prisma.doctorWorkerAssignment.findFirst({
      where: { workerId, active: true, hospitalId: worker.hospitalId! },
      include: { doctor: true },
    });
    return this.withAssignment(this.toResponse(worker), assignment);
  }

  async getWorkerByCode(hospitalUserId: bigint, workerCode: string) {
    const hospital = await this.getHospitalByUserId(hospitalUserId);
    const worker = await this.prisma.worker.findFirst({ where: { workerCode, hospitalId: hospital.id } });
    if (!worker) throw new NotFoundException('Worker not found');
    return this.getWorker(hospitalUserId, worker.id);
  }

  async getWorkerForDoctor(doctorUserId: bigint, workerId: bigint) {
    const doctor = await this.prisma.doctor.findUnique({ where: { userId: doctorUserId } });
    if (!doctor) throw new NotFoundException('Doctor profile not found');
    const assignment = await this.prisma.doctorWorkerAssignment.findFirst({ where: { doctorId: doctor.id, workerId, active: true }, include: { doctor: true } });
    if (!assignment) throw new NotFoundException('Worker is not assigned to you');
    const worker = await this.prisma.worker.findUnique({ where: { id: workerId } });
    if (!worker) throw new NotFoundException('Worker not found');
    return this.withAssignment(this.toResponse(worker), assignment);
  }

  async updateWorker(hospitalUserId: bigint, workerId: bigint, request: UpdateWorkerDto) {
    await this.getOwnedWorker(hospitalUserId, workerId);
    const updated = await this.prisma.worker.update({
      where: { id: workerId },
      data: { ...request, dateOfBirth: request.dateOfBirth ? new Date(request.dateOfBirth) : undefined },
    });
    return this.toResponse(updated);
  }

  async deactivateWorker(hospitalUserId: bigint, workerId: bigint) {
    await this.getOwnedWorker(hospitalUserId, workerId);
    const updated = await this.prisma.$transaction(async (tx) => {
      const worker = await tx.worker.findUnique({ where: { id: workerId } });
      if (!worker) throw new NotFoundException('Worker not found');
      await tx.doctorWorkerAssignment.updateMany({ where: { workerId, active: true }, data: { active: false, endedAt: new Date() } });
      await tx.workerQrCode.updateMany({ where: { workerId, status: 'ACTIVE' }, data: { status: 'REVOKED', revokedAt: new Date() } });
      if (worker.userId) await tx.user.update({ where: { id: worker.userId }, data: { status: 'INACTIVE' } });
      return tx.worker.update({ where: { id: workerId }, data: { active: false } });
    });
    return this.toResponse(updated);
  }

  async activateWorker(hospitalUserId: bigint, workerId: bigint) {
    await this.getOwnedWorker(hospitalUserId, workerId);
    const updated = await this.prisma.$transaction(async (tx) => {
      const worker = await tx.worker.update({ where: { id: workerId }, data: { active: true } });
      if (worker.userId) await tx.user.update({ where: { id: worker.userId }, data: { status: 'ACTIVE' } });
      return worker;
    });
    return this.toResponse(updated);
  }

  async getMyProfile(userId: bigint) {
    const worker = await this.prisma.worker.findUnique({
      where: { userId },
      include: { hospital: true, assignments: { where: { active: true }, include: { doctor: true, hospital: true }, take: 1 } },
    });
    if (!worker) throw new NotFoundException('Worker profile not found');
    return {
      ...this.toResponse(worker),
      hospital: worker.hospital ? { id: Number(worker.hospital.id), code: worker.hospital.hospitalCode, name: worker.hospital.name, city: worker.hospital.city, district: worker.hospital.district } : null,
      assignedDoctor: worker.assignments[0] ? { id: Number(worker.assignments[0].doctor.id), name: worker.assignments[0].doctor.fullName, specialization: worker.assignments[0].doctor.specialization, assignedAt: worker.assignments[0].assignedAt } : null,
    };
  }

  async updateMyProfile(userId: bigint, request: UpdateWorkerDto) {
    const worker = await this.prisma.worker.findUnique({ where: { userId } });
    if (!worker) throw new NotFoundException('Worker profile not found');
    if (!worker.active) throw new ForbiddenException('Worker account is inactive');
    const updated = await this.prisma.worker.update({ where: { id: worker.id }, data: { ...request, dateOfBirth: request.dateOfBirth ? new Date(request.dateOfBirth) : undefined } });
    return this.toResponse(updated);
  }

  async getMyMedicalRecords(userId: bigint) {
    const worker = await this.prisma.worker.findUnique({ where: { userId } });
    if (!worker) throw new NotFoundException('Worker profile not found');
    const records = await this.prisma.medicalRecord.findMany({ where: { workerId: worker.id }, orderBy: { visitDate: 'desc' }, include: { hospital: true, doctor: true, prescriptions: true, attachments: { where: { status: 'ACTIVE' } } } });
    return records.map((record) => ({
      id: Number(record.id), visitDate: record.visitDate, hospitalName: record.hospital.name, doctorName: record.doctor.fullName,
      symptoms: record.symptoms, diagnosis: record.diagnosis, treatment: record.treatment, notes: record.notes, updatedAt: record.updatedAt,
      prescriptions: record.prescriptions.map((p) => ({ id: Number(p.id), medicineName: p.medicineName, dosage: p.dosage, frequency: p.frequency, duration: p.duration, instructions: p.instructions })),
      attachments: record.attachments.map((a) => ({ id: Number(a.id), fileName: a.fileName, mimeType: a.mimeType, fileSize: a.fileSize, status: a.status, createdAt: a.createdAt })),
    }));
  }

  private withAssignment(response: any, assignment: any) {
    if (!assignment) return { ...response, assignedDoctor: null, assignedDoctorId: null, assignedDoctorName: null, assignedDoctorSpecialization: null };
    return {
      ...response,
      assignedDoctor: { id: Number(assignment.doctor.id), name: assignment.doctor.fullName, specialization: assignment.doctor.specialization, assignedAt: assignment.assignedAt },
      assignedDoctorId: Number(assignment.doctor.id), assignedDoctorName: assignment.doctor.fullName, assignedDoctorSpecialization: assignment.doctor.specialization,
    };
  }

  private toResponse(worker: any) {
    return {
      id: Number(worker.id), workerCode: worker.workerCode, fullName: worker.fullName, dateOfBirth: worker.dateOfBirth,
      gender: worker.gender, bloodGroup: worker.bloodGroup, phone: worker.phone, address: worker.address,
      emergencyContactName: worker.emergencyContactName, emergencyContactPhone: worker.emergencyContactPhone,
      emergencyContactRelation: worker.emergencyContactRelation, active: worker.active, createdAt: worker.createdAt,
    };
  }
}
